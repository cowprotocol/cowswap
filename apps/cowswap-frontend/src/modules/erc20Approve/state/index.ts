export * from './useUpdateTradeApproveState'
export * from './useGetUserApproveAmountState'
export * from './useSetUserApproveAmountModalState'
export * from './useIsPartialApproveSelectedByUser'
export * from './customApproveAmountInputState'
export * from './pendingApproveAmountModalState'
