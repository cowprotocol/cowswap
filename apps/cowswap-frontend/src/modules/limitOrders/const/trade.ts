import { Percent } from '@uniswap/sdk-core'

export const LIMIT_ORDER_SLIPPAGE = new Percent(0)
export const LOW_RATE_THRESHOLD_PERCENT = -5
