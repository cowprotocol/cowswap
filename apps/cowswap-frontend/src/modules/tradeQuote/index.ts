export * from './updaters/TradeQuoteUpdater'
export * from './state/tradeQuoteAtom'
export * from './hooks/useTradeQuote'
export * from './hooks/useSetTradeQuoteParams'
export * from './hooks/useQuoteParams'
export * from './hooks/useTradeQuoteFeeFiatAmount'
export * from './utils/quoteDeadline'
