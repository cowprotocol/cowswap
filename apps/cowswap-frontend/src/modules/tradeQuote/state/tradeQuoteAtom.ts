import { atom } from 'jotai'
import { atomWithReset } from 'jotai/utils'

import { OrderQuoteResponse } from '@cowprotocol/cow-sdk'

import type { LegacyFeeQuoteParams } from 'legacy/state/price/types'

import GpQuoteError from 'api/gnosisProtocol/errors/QuoteError'

export const DEFAULT_TRADE_QUOTE_STATE = {
  response: null,
  error: null,
  isLoading: false,
  quoteParams: null,
  localQuoteTimestamp: null,
}
export interface TradeQuoteState {
  response: OrderQuoteResponse | null
  error: GpQuoteError | null
  isLoading: boolean
  quoteParams: LegacyFeeQuoteParams | null
  localQuoteTimestamp: number | null
}

export const tradeQuoteAtom = atomWithReset<TradeQuoteState>(DEFAULT_TRADE_QUOTE_STATE)

export const updateTradeQuoteAtom = atom(null, (get, set, nextState: Partial<TradeQuoteState>) => {
  set(tradeQuoteAtom, () => {
    const prevState = get(tradeQuoteAtom)

    return {
      ...prevState,
      ...nextState,
      quoteParams: typeof nextState.quoteParams === 'undefined' ? prevState.quoteParams : nextState.quoteParams,
      localQuoteTimestamp: nextState.response ? Math.ceil(Date.now() / 1000) : null,
    }
  })
})
