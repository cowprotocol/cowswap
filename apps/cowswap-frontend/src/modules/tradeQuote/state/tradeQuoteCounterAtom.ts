import { atomWithReset } from 'jotai/utils'

import { QUOTE_POLLING_INTERVAL } from '../consts'

export const tradeQuoteCounterAtom = atomWithReset(QUOTE_POLLING_INTERVAL)
