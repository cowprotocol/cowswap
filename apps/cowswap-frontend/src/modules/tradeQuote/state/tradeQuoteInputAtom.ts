import { atom } from 'jotai'

import { Currency, CurrencyAmount } from '@uniswap/sdk-core'

export type SellTokenAddress = string

export interface TradeQuoteInputState {
  amount: CurrencyAmount<Currency> | null
  fastQuote?: boolean
  partiallyFillable?: boolean
}

export const tradeQuoteInputAtom = atom<TradeQuoteInputState>({ amount: null })
