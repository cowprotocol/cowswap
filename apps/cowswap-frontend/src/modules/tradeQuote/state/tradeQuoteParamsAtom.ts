import { atom } from 'jotai'

import { Currency, CurrencyAmount } from '@uniswap/sdk-core'

export interface TradeQuoteParamsState {
  amount: CurrencyAmount<Currency> | null
  fastQuote?: boolean
}

export const tradeQuoteParamsAtom = atom<TradeQuoteParamsState>({ amount: null })
