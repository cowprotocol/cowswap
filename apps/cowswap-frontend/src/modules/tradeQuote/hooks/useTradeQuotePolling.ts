import { useAtomValue, useSetAtom } from 'jotai'
import { useLayoutEffect, useMemo } from 'react'

import { useDebounce } from '@cowprotocol/common-hooks'
import { onlyResolvesLast } from '@cowprotocol/common-utils'
import { OrderQuoteResponse } from '@cowprotocol/cow-sdk'
import { useAreUnsupportedTokens } from '@cowprotocol/tokens'

import ms from 'ms.macro'

import { useUpdateCurrencyAmount } from 'modules/trade/hooks/useUpdateCurrencyAmount'
import { updateTradeQuoteAtom } from 'modules/tradeQuote/state/tradeQuoteAtom'

import { getQuote } from 'api/gnosisProtocol/api'
import GpQuoteError, { GpQuoteErrorCodes } from 'api/gnosisProtocol/errors/QuoteError'

import { useProcessUnsupportedTokenError } from './useProcessUnsupportedTokenError'
import { useQuoteParams } from './useQuoteParams'

import { tradeQuoteParamsAtom } from '../state/tradeQuoteParamsAtom'

export const PRICE_UPDATE_INTERVAL = ms`30s`
const AMOUNT_CHANGE_DEBOUNCE_TIME = ms`300`

// Solves the problem of multiple requests
const getQuoteOnlyResolveLast = onlyResolvesLast<OrderQuoteResponse>(getQuote)

export function useTradeQuotePolling() {
  const { amount } = useAtomValue(tradeQuoteParamsAtom)
  const amountStr = useDebounce(
    useMemo(() => amount?.quotient.toString() || null, [amount]),
    AMOUNT_CHANGE_DEBOUNCE_TIME
  )
  const quoteParams = useQuoteParams(amountStr)

  const updateQuoteState = useSetAtom(updateTradeQuoteAtom)
  const updateCurrencyAmount = useUpdateCurrencyAmount()
  const getIsUnsupportedTokens = useAreUnsupportedTokens()
  const processUnsupportedTokenError = useProcessUnsupportedTokenError()

  useLayoutEffect(() => {
    if (!quoteParams) {
      updateQuoteState({ response: null, isLoading: false })
      return
    }

    const isUnsupportedTokens = getIsUnsupportedTokens(quoteParams)

    // Don't fetch quote if token is not supported
    if (isUnsupportedTokens) {
      return
    }

    const fetchQuote = () => {
      updateQuoteState({ isLoading: true })

      getQuoteOnlyResolveLast(quoteParams)
        .then((response) => {
          const { cancelled, data } = response

          if (cancelled) {
            return
          }

          updateQuoteState({ response: data, quoteParams, isLoading: false, error: null })
        })
        .catch((error: GpQuoteError) => {
          console.log('[useGetQuote]:: fetchQuote error', error)
          updateQuoteState({ isLoading: false, error })

          if (error.type === GpQuoteErrorCodes.UnsupportedToken) {
            processUnsupportedTokenError(error, quoteParams)
          }
        })
    }

    fetchQuote()

    const intervalId = setInterval(fetchQuote, PRICE_UPDATE_INTERVAL)

    return () => clearInterval(intervalId)
  }, [quoteParams, updateQuoteState, updateCurrencyAmount, processUnsupportedTokenError, getIsUnsupportedTokens])

  return null
}
