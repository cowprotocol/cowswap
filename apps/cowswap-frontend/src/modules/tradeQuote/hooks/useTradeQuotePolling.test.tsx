import { ReactNode } from 'react'

import { COW, WETH_SEPOLIA, ZERO_ADDRESS } from '@cowprotocol/common-const'
import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { WalletInfo, walletInfoAtom } from '@cowprotocol/wallet'
import { CurrencyAmount } from '@uniswap/sdk-core'

import { renderHook } from '@testing-library/react-hooks'
import { orderBookApi } from 'cowSdk'
import { JotaiTestProvider, WithMockedWeb3 } from 'test-utils'

import { LimitOrdersDerivedState, limitOrdersDerivedStateAtom } from 'modules/limitOrders/state/limitOrdersRawStateAtom'
import * as tokensModule from 'modules/tokens'
import { DEFAULT_TRADE_DERIVED_STATE } from 'modules/trade'

import { useTradeQuotePolling } from './useTradeQuotePolling'

import { tradeQuoteParamsAtom } from '../state/tradeQuoteParamsAtom'

jest.mock('modules/zeroApproval/hooks/useZeroApprovalState')
jest.mock('common/hooks/useGetMarketDimension')

const getQuoteMock = jest.spyOn(orderBookApi, 'getQuote')
const useEnoughBalanceAndAllowanceMock = jest.spyOn(tokensModule, 'useEnoughBalanceAndAllowance')

const inputCurrencyAmount = CurrencyAmount.fromRawAmount(WETH_SEPOLIA, 10_000_000)
const outputCurrencyAmount = CurrencyAmount.fromRawAmount(COW[SupportedChainId.SEPOLIA], 2_000_000)

const walletInfoMock: WalletInfo = {
  chainId: 1,
  account: '0x333333f332a06ecb5d20d35da44ba07986d6e203',
  active: true,
}

const limitOrdersDerivedStateMock: LimitOrdersDerivedState = {
  ...DEFAULT_TRADE_DERIVED_STATE,
  inputCurrency: inputCurrencyAmount.currency,
  outputCurrency: outputCurrencyAmount.currency,
  inputCurrencyAmount,
  outputCurrencyAmount,
  isUnlocked: true,
}

const jotaiMock = [
  [tradeQuoteParamsAtom, { amount: inputCurrencyAmount }],
  [limitOrdersDerivedStateAtom, limitOrdersDerivedStateMock],
]

const Wrapper =
  (mocks: any) =>
  ({ children }: { children: ReactNode }) =>
    (
      <WithMockedWeb3 location={{ pathname: '/5/limit' }}>
        <JotaiTestProvider initialValues={mocks}>{children}</JotaiTestProvider>
      </WithMockedWeb3>
    )

describe('useTradeQuotePolling()', () => {
  beforeEach(() => {
    jest.clearAllMocks()

    getQuoteMock.mockImplementation(() => new Promise(() => void 0))
    useEnoughBalanceAndAllowanceMock.mockReturnValue({ enoughBalance: true, enoughAllowance: true })
  })

  describe('When wallet is connected', () => {
    it('Then should put account address into "useAddress" field in the quote request', () => {
      // Arrange
      const mocks = [...jotaiMock, [walletInfoAtom, walletInfoMock]]

      // Act
      renderHook(() => useTradeQuotePolling(), { wrapper: Wrapper(mocks) })

      // Assert
      const callParams = getQuoteMock.mock.calls[0][0]

      expect(callParams.from).toBe(walletInfoMock.account) // useAddress field value
      expect(getQuoteMock).toHaveBeenCalledTimes(1)
      expect(callParams).toMatchSnapshot()
    })
  })

  describe('When wallet is NOT connected', () => {
    it('Then the "useAddress" field in the quote request should be 0x000...0000', () => {
      // Arrange
      const mocks = [...jotaiMock, [walletInfoAtom, { ...walletInfoMock, account: undefined }]]

      // Act
      renderHook(() => useTradeQuotePolling(), { wrapper: Wrapper(mocks) })

      // Assert
      const callParams = getQuoteMock.mock.calls[0][0]

      expect(callParams.from).toBe(ZERO_ADDRESS) // useAddress field value
      expect(getQuoteMock).toHaveBeenCalledTimes(1)
      expect(callParams).toMatchSnapshot()
    })
  })
})
