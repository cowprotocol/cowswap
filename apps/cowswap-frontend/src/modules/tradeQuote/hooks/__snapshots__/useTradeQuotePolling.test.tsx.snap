// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTradeQuotePolling() When wallet is NOT connected Then the "useAddress" field in the quote request should be 0x000...0000 1`] = `
{
  "appData": "{"version":"1.1.0","appCode":"CoW Swap","metadata":{}}",
  "appDataHash": "0x3f13237929334d44bdb32b189d1644965915936e62004fb77ec7bc861cddc7a4",
  "buyToken": "0x0625aFB445C3B6B7B929342a04A22599fd5dBB59",
  "from": "0x0000000000000000000000000000000000000000",
  "kind": "sell",
  "partiallyFillable": false,
  "priceQuality": "optimal",
  "receiver": "0x0000000000000000000000000000000000000000",
  "sellAmountBeforeFee": "10000000",
  "sellToken": "0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14",
  "validTo": undefined,
}
`;

exports[`useTradeQuotePolling() When wallet is connected Then should put account address into "useAddress" field in the quote request 1`] = `
{
  "appData": "{"version":"1.1.0","appCode":"CoW Swap","metadata":{}}",
  "appDataHash": "0x3f13237929334d44bdb32b189d1644965915936e62004fb77ec7bc861cddc7a4",
  "buyToken": "0x0625aFB445C3B6B7B929342a04A22599fd5dBB59",
  "from": "0x333333f332a06ecb5d20d35da44ba07986d6e203",
  "kind": "sell",
  "partiallyFillable": false,
  "priceQuality": "optimal",
  "receiver": "0x333333f332a06ecb5d20d35da44ba07986d6e203",
  "sellAmountBeforeFee": "10000000",
  "sellToken": "0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14",
  "validTo": undefined,
}
`;
